### DISCUSSION TITLE
From D dimensions to 4

### DISCUSSION THREAD

**Discussion by alaban92** [FeynCalc/feyncalc#133] [General]
Hi FeynCalc experts,

I'm trying to implement my old FenyCalc codes to version 9.3.1 of FeynCalc in Mathematica 10.2. I had no problems with the command Oneloop[] when I went from D dimension to 4 dimensions, so let's say that I have my amplitude ampt1

amp1a = Calc[ampt1] // FCE;

FCSetDiracGammaScheme["BMHV"] (because there are gamma5 in my amplitude)

amp2a = DiracTrace[amp1a] // DiracSimplify;

amp3a = TID[amp2a, k,ToPaVe -> True]; 

I used to replace D=4+2epsilon and expand around epsilon to obtain my 4 dimensions results. Unfortunately, I can't reproduce my old result now. I read https://github.com/FeynCalc/feyncalc/discussions/66 and tried with the option PaVeLimitTo4->True in the dv version but it isn't working as I obtain that my amplitude is of the form 1/0.

Thanks for your time and your help.

**Comment by vsht** [comment]
Hi,

[PaVeLimitTo4](https://feyncalc.github.io/FeynCalcBookDev/PaVeLimitTo4.html) always works with `(D-4)`, so it should not matter whether you use `4+2 ep`, `4-2 ep`, `4- ep` or `4+ep`. 

As to the 1/0 error, I'd need an explicit example to understand what goes wrong here.

**Reply by alaban92** [reply]
Hi,

here is my example

FeynCalc 9.3.1 (development version). For help, use the documentation center, check out the wiki or visit the forum.......


{SP[p1, p1] = mz^2, SP[p2, p2] = 0, SP[p1, p2] = (-mz^2 + Q^2)/2, 
  SP[p1, q] = (Q^2 + mz^2)/2, SP[p2, q] = (Q^2 - mz^2)/2, 
  SP[q, q] = Q^2};


(*-----------------------*)

ampt1 = ((GSD[k] + m).(GAD[\[Mu]]).(gv - ga GA5).(GSD[k] - GSD[q] + 
      m).GAD[\[Beta]].(GSD[k] - GSD[p1] + m).GAD[\[Alpha]].(gv - 
      ga GA5)) FAD[{k, m}, {k - p1, m}, {k - q, m}]



(*-----------------------*)

amp1a = Calc[ampt1] // FCE;

FCSetDiracGammaScheme["BMHV"]

amp2a = DiracTrace[amp1a] // DiracSimplify // Simplify // FCE;



amp3a = TID[amp2a, k, ToPaVe -> True, PaVeLimitTo4 -> True] // FCE;

Power::infy: Infinite expression 1/0^2 encountered. >>

Power::infy: Infinite expression 1/0^2 encountered. >>

Power::infy: Infinite expression 1/0^2 encountered. >>

General::stop: Further output of Power::infy will be suppressed during this calculation. >>

PaVeLimitTo4::failmsg: Error! PaVeLimitTo4 has encountered a fatal problem and must abort the computation. The problem reads: Cannot handle PaVe functions multiplied by 1/Epsilon poles.


sorry for the late reply

**Comment by vsht** [comment]
Hi,

you need to set also
```
{SPD[p1, p1] = mz^2, SPD[p2, p2] = 0, SPD[p1, p2] = (-mz^2 + Q^2)/2,
SPD[p1, q] = (Q^2 + mz^2)/2, SPD[p2, q] = (Q^2 - mz^2)/2,
SPD[q, q] = Q^2};
```
since the scalar products are `D`-dimensional during the reduction. Or just use `ScalarProduct[a,b] = ...` which sets
both `D` and `4` dimensional values.

Otherwise the amplitude after the reduction contains kinematic singularities that blow up when `PaVeLimitTo4` sets the dimension
to 4, so that your initial scalar product definitions are applied.
